File: application/app/domain/validation/validators/ValidRouteConfigurationValidator.scala
package domain.validation.validators

import domain.model.{EndpointConfiguration, RouteType}
import domain.validation.constraints.ValidRouteConfiguration
import javax.validation.{ConstraintValidator, ConstraintValidatorContext}

class ValidRouteConfigurationValidator extends ConstraintValidator[ValidRouteConfiguration, EndpointConfiguration] {
  override def isValid(configuration: EndpointConfiguration, context: ConstraintValidatorContext): Boolean = {
    if (configuration == null) {
      return true
    }
    if (configuration.routes == null) {
      return true
    }

    checkRouteNumberByTpes(context, configuration, RouteType.CONNECT) &&
      checkRouteNumberByTpes(context, configuration, RouteType.DISCONNECT) &&
      checkRouteNumberByTpes(context, configuration, RouteType.DEFAULT)
  }

  private def checkRouteNumberByTpes(context: ConstraintValidatorContext, configuration: EndpointConfiguration, routeType: RouteType.Value): Boolean = {
    val routes = configuration.getRoutes(routeType)
    if (routes.size > 1) {
      context.disableDefaultConstraintViolation();
      context.buildConstraintViolationWithTemplate(s"Endpoint can have at most 1 route of type $routeType")
        .addConstraintViolation()
      return false
    }
    true
  }
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/validation/validators/ValidRouteConfigurationValidator.java
package com.cosmin.wsgateway.domain.validation.validators;

import com.cosmin.wsgateway.domain.EndpointConfiguration;
import com.cosmin.wsgateway.domain.Route;
import com.cosmin.wsgateway.domain.validation.constraints.ValidRouteConfiguration;
import java.util.Set;
import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class ValidRouteConfigurationValidator implements
        ConstraintValidator<ValidRouteConfiguration, EndpointConfiguration> {

    @Override
    public boolean isValid(EndpointConfiguration value, ConstraintValidatorContext context) {
        if (value == null) {
            return true;
        }
        if (value.getRoutes() == null) {
            return true;
        }

        return checkRouteNumberByTypes(context, value, Route.Type.CONNECT)
                && checkRouteNumberByTypes(context, value, Route.Type.DEFAULT)
                && checkRouteNumberByTypes(context, value, Route.Type.DISCONNECT);
    }

    private Boolean checkRouteNumberByTypes(
            ConstraintValidatorContext context, EndpointConfiguration configuration, Route.Type type
    ) {
        Set<Route> routes = configuration.getRoutesByType(type);
        if (routes.size() > 1) {
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate("Endpoint can have at most 1 route of type " + type)
                    .addConstraintViolation();
            return false;
        }

        return true;
    }
}
